app/services/attachments/finish_direct_upload_service.rb

```diff
@@ -0,0 +1,142 @@
+# frozen_string_literal: true
+
+# -- copyright
+# OpenProject is an open source project management software.
+# Copyright (C) the OpenProject GmbH
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License version 3.
+#
+# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
+# Copyright (C) 2006-2013 Jean-Philippe Lang
+# Copyright (C) 2010-2013 the ChiliProject Team
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+# ++
+
+module Attachments
+  class FinishDirectUploadService < BaseServices::BaseContracted
+    def initialize(user:, model:, contract_class: nil, contract_options: {})
+      self.model = model
+      super(user:, contract_class:, contract_options:)
+    end
+
+    protected
+
+    alias_method :attachment, :model
+
+    def service_context(send_notifications:, &)
+      # Overwriting the call to in_context to place the semaphore on the container and not on the attachment.
+      # Since the service will write a journal on the container, there should not be another
+      # process that modifies the container while this service is running.
+      in_context(attachment.container, send_notifications:, &)
+    end
+
+    def validate_params(_params)
+      super.tap do |call|
+        validate_local_file_exists(call)
+      end
+    end
+
+    def before_perform(*)
+      super.tap do
+        set_attachment_parameters
+      end
+    end
+
+    def persist(call)
+      super.tap do
+        attachment.save!
+      end
+    end
+
+    def after_perform(service_call)
+      super.tap do
+        journalize_container
+        attachment_created_event
+        schedule_jobs
+      end
+    end
+
+    def validate_local_file_exists(call)
+      unless local_file
+        call.errors.add(:base, "File for attachment #{attachment.filename} was not uploaded.")
+        call.success = false
+      end
+    end
+
+    def set_attachment_parameters
+      attachment.extend(OpenProject::ChangedBySystem)
+      attachment.change_by_system do
+        attachment.status = :uploaded
+        attachment.set_file_size local_file
+        attachment.set_content_type local_file
+        attachment.set_digest local_file
```

Comment:
Won't following do the same? Also I'm not sure it is needed at all, wouldn't it be already called before?
```suggestion
        attachment.file = local_file
```

Reply:
You are right, I just c&p-ed it over. But this of course reads nicer.